### PR TITLE
mfcj430w-cupswrapper: init at 3.0.0-1

### PR DESCRIPTION
**Pull request by svavs**
###### Motivation for this change
 The printer Brother MFC-J430W is not recognized by default (gutenprint, brgenml1cupswrapper, brgenml1lpr). Requires official drivers.

###### Things done
This change allows to install the official Brother driver and CUPS wrapper through services.printing.drivers.

- [x ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ x] Ensured that relevant documentation is up to date
- [ x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 123305` at 32bc98b1 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>mfcj430w-cupswrapper</li>
    <li>mfcj430wlpr</li>
  </ul>
</details>
<details>
  <summary>7 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/stale-substitute.md">stale-substitute</a>

Stale substituteInPlace detected.
substituteStream(): WARNING: pattern '`mv ' doesn't match anything in file 'cupswrapper/cupswrappermfcj430w'
Near pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:85:0:
```
   |
85 |     description = "Brother MFC-J430W CUPS wrapper driver";
   | ^
```
</li>
    <li>warning: unused-argument

Unused argument: `pkgs`.
Near pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:30:79:
```
   |
30 | { lib, stdenv, fetchurl, makeWrapper, cups, perl, coreutils, gnused, gnugrep, pkgs, mfcj430wlpr }:
   |                                                                               ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/patch-phase.md">patch-phase</a>

`patchPhase` should not be overridden, use `postPatch` instead.

Near pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:44:3:
```
   |
44 |   patchPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/stale-substitute.md">stale-substitute</a>

Stale substituteInPlace detected.
substituteStream(): WARNING: pattern '`cp ' doesn't match anything in file 'cupswrapper/cupswrappermfcj430w'
Near pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:85:0:
```
   |
85 |     description = "Brother MFC-J430W CUPS wrapper driver";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:63:3:
```
   |
63 |   installPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:57:3:
```
   |
57 |   buildPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl2` is a deprecated license, please check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:86:5:
```
   |
86 |     license = lib.licenses.gpl2;
   |     ^
```
</li>
  </ul>
</details>

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -9617,6 +9617,12 @@
     githubId = 144952;
     name = "Suvash Thapaliya";
   };
+  svavs = {
+    name = "Silvano Sallese";
+    email = "silvano.sallese+nixpkgs@gmail.com";
+    github = "svavs";
+    githubId = 1369198;
+  };
```
Comment (legendofmiracles) [inline maintainers/maintainer-list.nix:9625]: Separate commit: `maintainers: add svavs`

File: pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  meta = {
+    homepage = "http://www.brother.com/";
+    description = "Brother MFC-J430W CUPS wrapper driver";
+    license = lib.licenses.gpl2Only;
+    platforms = lib.platforms.linux;
+    downloadPage = "https://support.brother.com/g/b/downloadlist.aspx?c=us&lang=en&prod=mfcj430w_all&os=128&autolayerclose=1";
+    maintainers = with lib.maintainers; [ svavs ];
+  };
+}
```
Comment (legendofmiracles) [inline pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:91]:
```suggestion
  meta = with lib; {
    homepage = "http://www.brother.com/";
    description = "Brother MFC-J430W CUPS wrapper driver";
    license = licenses.gpl2Only;
    platforms = platforms.linux;
    downloadPage = "https://support.brother.com/g/b/downloadlist.aspx?c=us&lang=en&prod=mfcj430w_all&os=128&autolayerclose=1";
    maintainers = with maintainers; [ svavs ];
  };
}
```

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (svavs) [reply maintainers/maintainer-list.nix:9625]: Done. Thank you.

File: pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix (hunk 2 above)
Comment (svavs) [reply pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix]: I've not directly commit your suggestion, sorry, I preferred to do a single commit for both files. Thank you.

File: pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    cp ppd/brother_mfcj430w_printer_en.ppd $TARGETFOLDER/
+
+    mkdir -p $CUPSPPD_DIR
+    ln -s $TARGETFOLDER/brother_mfcj430w_printer_en.ppd $CUPSPPD_DIR
+    '';
+
+  cleanPhase = ''
+    cd brcupsconfig
+    make clean
+    '';
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:81]:
```suggestion
  '';
```

File: pkgs/misc/cups/drivers/mfcj430wlpr/default.nix [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+    mkdir -p $out/lib/cups/filter/
+    ln -s $out/opt/brother/Printers/mfcj430w/lpd/filtermfcj430w $out/lib/cups/filter/brother_lpdwrapper_mfcj430w
+
+    wrapProgram $out/opt/brother/Printers/mfcj430w/lpd/psconvertij2 \
+    --prefix PATH ":" ${ lib.makeBinPath [ gnused coreutils gawk ] }
+
+    wrapProgram $out/opt/brother/Printers/mfcj430w/lpd/filtermfcj430w \
+    --prefix PATH ":" ${ lib.makeBinPath [ ghostscript a2ps file gnused coreutils ] }
+    '';
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/mfcj430wlpr/default.nix:65]:
```suggestion
  '';
```

File: pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    substituteInPlace $WRAPPER \
+    --replace "\`cp " "\`cp -p " \
+    --replace "\`mv " "\`cp -p "
+    '';
+
+  buildPhase = ''
+    cd brcupsconfig
+    make all
+    cd ..
+    '';
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:61]:
```suggestion
  '';
```

File: pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+    mkdir -p $TARGETFOLDER
+
+    cp brcupsconfig/brcupsconfpt1 $TARGETFOLDER
+    cp cupswrapper/cupswrappermfcj430w $TARGETFOLDER/
+    cp ppd/brother_mfcj430w_printer_en.ppd $TARGETFOLDER/
+
+    mkdir -p $CUPSPPD_DIR
+    ln -s $TARGETFOLDER/brother_mfcj430w_printer_en.ppd $CUPSPPD_DIR
+    '';
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:76]:
```suggestion
  '';
```

File: pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+    substituteInPlace $WRAPPER \
+    --replace /opt "${mfcj430wlpr}/opt" \
+    --replace /usr "${mfcj430wlpr}/usr" \
+    --replace /etc "$out/etc"
+
+    substituteInPlace $WRAPPER \
+    --replace "\`cp " "\`cp -p " \
+    --replace "\`mv " "\`cp -p "
+    '';
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:55]:
```suggestion
  '';
```

File: pkgs/misc/cups/drivers/mfcj430wlpr/default.nix [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+  meta = with lib; {
+    homepage = "http://www.brother.com/";
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/mfcj430wlpr/default.nix:68]:
```suggestion
    homepage = "https://www.brother.com/";
```

File: pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    ln -s $TARGETFOLDER/brother_mfcj430w_printer_en.ppd $CUPSPPD_DIR
+    '';
+
+  cleanPhase = ''
+    cd brcupsconfig
+    make clean
+    '';
+
+  meta = with lib; {
+    homepage = "http://www.brother.com/";
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:84]:
```suggestion
    homepage = "https://www.brother.com/";
```

File: pkgs/misc/cups/drivers/mfcj430wlpr/default.nix [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    patchelf --set-interpreter ${stdenv.glibc.out}/lib/ld-linux.so.2 $out/opt/brother/Printers/mfcj430w/lpd/brmfcj430wfilter
+
+    mkdir -p $out/lib/cups/filter/
+    ln -s $out/opt/brother/Printers/mfcj430w/lpd/filtermfcj430w $out/lib/cups/filter/brother_lpdwrapper_mfcj430w
+
+    wrapProgram $out/opt/brother/Printers/mfcj430w/lpd/psconvertij2 \
+    --prefix PATH ":" ${ lib.makeBinPath [ gnused coreutils gawk ] }
+
+    wrapProgram $out/opt/brother/Printers/mfcj430w/lpd/filtermfcj430w \
+    --prefix PATH ":" ${ lib.makeBinPath [ ghostscript a2ps file gnused coreutils ] }
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/mfcj430wlpr/default.nix:64]:
```suggestion
      --prefix PATH ":" ${lib.makeBinPath [ gnused coreutils gawk ]}

    wrapProgram $out/opt/brother/Printers/mfcj430w/lpd/filtermfcj430w \
      --prefix PATH ":" ${lib.makeBinPath [ ghostscript a2ps file gnused coreutils ]}
```

File: pkgs/misc/cups/drivers/mfcj430wlpr/default.nix [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ cups ghostscript dpkg a2ps ];
+
+  dontUnpack = true;
+
+  installPhase = ''
+    dpkg-deb -x $src $out
+
+    substituteInPlace $out/opt/brother/Printers/mfcj430w/lpd/filtermfcj430w \
+    --replace /opt "$out/opt" \
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/mfcj430wlpr/default.nix:51]:
```suggestion
      --replace /opt "$out/opt" \
```

File: pkgs/misc/cups/drivers/mfcj430wlpr/default.nix [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    sha256 = "0an3xq9gmml2dxwba6jk6m2aypaa8jii7lzqgk7y46acf5ar09da";
+  };
+
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ cups ghostscript dpkg a2ps ];
+
+  dontUnpack = true;
+
+  installPhase = ''
+    dpkg-deb -x $src $out
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/mfcj430wlpr/default.nix:48]: This belongs into unpackCmd

File: pkgs/misc/cups/drivers/mfcj430wlpr/default.nix [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+#     URI example
+#     ~  `lpd://BRW0080927AFBCE/binary_p1`
+#     ~  `ipp://your-printer-ip-address:631/ipp`
+#
+#
+## Testing
+# . used on NixOS 20.09 with Cinnamon using system-config-printer
+#
+
+
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/mfcj430wlpr/default.nix:30]:
```suggestion
```

File: pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    CUPSFILTER_DIR=$out/lib/cups/filter
+
+    mkdir -p $TARGETFOLDER
+
+    cp brcupsconfig/brcupsconfpt1 $TARGETFOLDER
+    cp cupswrapper/cupswrappermfcj430w $TARGETFOLDER/
+    cp ppd/brother_mfcj430w_printer_en.ppd $TARGETFOLDER/
+
+    mkdir -p $CUPSPPD_DIR
+    ln -s $TARGETFOLDER/brother_mfcj430w_printer_en.ppd $CUPSPPD_DIR
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:75]:
```suggestion
    mkdir -p $out/opt/brother/Printers/mfcj430w/cupswrapper/ $out/share/cups/model
    cp brcupsconfig/brcupsconfpt1 cupswrapper/cupswrappermfcj430w ppd/brother_mfcj430w_printer_en.ppd $out/opt/brother/Printers/mfcj430w/cupswrapper/
    
    ln -s $TARGETFOLDER/brother_mfcj430w_printer_en.ppd $out/lib/cups/filter
```
Do we really need $out/opt?

File: pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+    substituteInPlace $WRAPPER \
+    --replace "\`cp " "\`cp -p " \
+    --replace "\`mv " "\`cp -p "
+  '';
+
+  buildPhase = ''
+    cd brcupsconfig
+    make all
+    cd ..
+  '';
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:61]: Please use makeTarget and pre and postBuild to change the directories.

File: pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+
+    substituteInPlace $WRAPPER \
+    --replace /opt "${mfcj430wlpr}/opt" \
+    --replace /usr "${mfcj430wlpr}/usr" \
+    --replace /etc "$out/etc"
+
+    substituteInPlace $WRAPPER \
+    --replace "\`cp " "\`cp -p " \
+    --replace "\`mv " "\`cp -p "
+  '';
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:55]:
```suggestion
  postPatch = ''
    substituteInPlace cupswrapper/cupswrappermfcj430w \
    --replace /opt "${mfcj430wlpr}/opt" \
    --replace /usr "${mfcj430wlpr}/usr" \
    --replace /etc "$out/etc" \
    --replace "\`cp " "\`cp -p " \
    --replace "\`mv " "\`cp -p "
  '';
```

File: pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+#     URI example
+#     ~  `lpd://BRW0080927AFBCE/binary_p1`
+#     ~  `ipp://your-printer-ip-address:631/ipp`
+#
+#
+## Testing
+# . used on NixOS 20.09 with Cinnamon using system-config-printer
+#
+
+
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/mfcj430wcupswrapper/default.nix:29]:
```suggestion
```

**Comment by Artturin** [comment]
Cleaned up a bit so anyone interested can pick up the PR

**Pull request closed without merging**